### ISSUE TITLE
Ellipsoid resulting model is bad

### ISSUE THREAD

**Issue by akenmorris** [SCIInstitute/ShapeWorks#1082]
Results are terrible in 6.0-RC5.  When did this happen?

6.0-RC5:

https://user-images.githubusercontent.com/1693349/109743941-bd834000-7b8e-11eb-8330-95b732494178.mp4

5.5:

https://user-images.githubusercontent.com/1693349/109743954-c2e08a80-7b8e-11eb-8a57-30a099a3651b.mp4

Compare the variance chart:

6.0-RC5:

<img width="1152" alt="6 0rc5-ellipsoid" src="https://user-images.githubusercontent.com/1693349/109743980-cbd15c00-7b8e-11eb-9c2d-94ed10a81a82.png">

5.5:

<img width="1292" alt="5 5-ellipsoid" src="https://user-images.githubusercontent.com/1693349/109743992-d4299700-7b8e-11eb-9193-4e610e71513a.png">

**Comment by akenmorris** [comment]
It occurs to me that the ellipsoid dataset has changed.  Can someone explain the difference between "v0" and "v1"?  ShapeWorks 5.5 doesn't perform any better on the current ellipsoid dataset/parameters.  @sheryjoe @jadie1 @iyerkrithika21

**Comment by iyerkrithika21** [comment]
@akenmorris  "v1" is the un-groomed ellipsoid datasets generated using the ShapeCohortGenPackage. It includes all the following modes of variations:
1. random radii
2. random orientation
3. some segmentations which touch boundaries 
4. randomized centers 

Whereas the old dataset had only one mode of variation, I suppose. @jadie1, correct me if I am wrong regarding v0.

**Comment by sheryjoe** [comment]
@jadie1 We agreed before that we will provide two ellipsoid datasets, one aligned (similar to v0) and one not aligned (to demonstrate grooming).  I think v1 is the latter.

**Comment by jadie1** [comment]
Yes ellipsoid-v1 is as @iyerkrithika21 described, so it should have more than one mode of variation. ellipsoid-v0 is the original data for which the ungroomed version is the same as the groomed. 
We also added ellipsoid_aligned-v1 which was generated using the cohort generation code as ellipsoid-v1 was but it is already aligned (center, orientation, and image size are the same). The difference between ellipsoid-v0 and ellipsoid_aligned-v1 is ellipsoid-v0 only has one mode of variation whereas ellipsoid_aligned-v1 varies along all radii.

**Comment by sheryjoe** [comment]
> Yes ellipsoid-v1 is as @iyerkrithika21 described, so it should have more than one mode of variation. ellipsoid-v0 is the original data for which the ungroomed version is the same as the groomed.

Why should v1 have more modes? The use case for this data should factor out modes related to misalignment.

> We also added ellipsoid_aligned-v1 which was generated using the cohort generation code as ellipsoid-v1 was but it is already aligned (center, orientation, and image size are the same). The difference between ellipsoid-v0 and ellipsoid_aligned-v1 is ellipsoid-v0 only has one mode of variation whereas ellipsoid_aligned-v1 varies along all radii.

This is becoming a bit confusing. Do we need to keep two aligned ellispoids data, one with one mode and the other with three modes? I think the aligned should only have one mode (pick one radius for variation) as this is supposed to be the simplest toy example for the optimization.

**Comment by akenmorris** [comment]
Regardless of alignment or number of modes of variation, this set of parameters shows poor correspondence and sampling:

<img width="1259" alt="image" src="https://user-images.githubusercontent.com/1693349/109833909-e6431e00-7bfe-11eb-832f-6eb755a136f0.png">

**Comment by jadie1** [comment]
> > Yes ellipsoid-v1 is as @iyerkrithika21 described, so it should have more than one mode of variation. ellipsoid-v0 is the original data for which the ungroomed version is the same as the groomed.
> 
> Why should v1 have more modes? The use case for this data should factor out modes related to misalignment.
>
Because the radius varies along all three axes there should be more modes of variation right?
> > We also added ellipsoid_aligned-v1 which was generated using the cohort generation code as ellipsoid-v1 was but it is already aligned (center, orientation, and image size are the same). The difference between ellipsoid-v0 and ellipsoid_aligned-v1 is ellipsoid-v0 only has one mode of variation whereas ellipsoid_aligned-v1 varies along all radii.
> 
> This is becoming a bit confusing. Do we need to keep two aligned ellispoids data, one with one mode and the other with three modes? I think the aligned should only have one mode (pick one radius for variation) as this is supposed to be the simplest toy example for the optimization.

I can remove ellipsoid_aligned-v1 or ellipsoid-v0. We don't have use cases that use both, I just left ellpsoid-v0 for users using older versions of ShapeWorks. This was the original request for creating ellpsoid_aligned-v1: "Can we regenerate v0 using the same code as v1 by disabling all transformations? This way we maintain consistency and have both segmentations and nice meshes to test both for fixed domains and cutting planes." 

I am not sure why the model has poor correspondence. I guess the parameters need to be retuned?  Should I try with new ellipsoids that only vary along one radius?

**Comment by akenmorris** [comment]
The grooming has problems as well:

<img width="655" alt="image" src="https://user-images.githubusercontent.com/1693349/110020763-2da4d980-7ce7-11eb-9290-55a1b63d303b.png">

Notice ellipsoid 9 is not even centered with the others.

**Comment by iyerkrithika21** [comment]
This is how the groomed ellipsoids look for me in windows and Linux. 
Ellipsoid 9 seems centered for me.

![image](https://user-images.githubusercontent.com/11266258/110027761-9d1ec700-7cef-11eb-8fa6-2c411d35586c.png)

**Comment by jadie1** [comment]
Looks centered for me too - just ran it again on the release_v6.0 branch
![image](https://user-images.githubusercontent.com/54382613/110028656-c2600500-7cf0-11eb-9672-29fca976f8a6.png)

**Comment by jadie1** [comment]
The correspondence points don't look great for me either though, maybe it needs more smoothing? To me it looks like there are four modes of variation - the first is size, and the next three are radii in the x, y, and z direction.

**Comment by akenmorris** [comment]
@jadie1, @iyerkrithika21, I just tried on a different Mac, fresh install of shape works 6.0 RC5, new conda installs, everything.  Look at the alignment for the samples (no need to even look at DTs):


https://user-images.githubusercontent.com/1693349/110035220-9c3e6300-7cf8-11eb-9cd0-9162212dd61c.mp4

**Comment by akenmorris** [comment]
Oh, hah, @jadie1 , @iyerkrithika21 , you must turn **off** the `center` checkbox in Studio or it will auto center them for you for display purposes.

**Comment by jadie1** [comment]
Oh yeah I see now. I just ran the getting-started-with-grooming-segmentations.ipynb which uses the same data and it seems to me it has the same problem but not as bad. See the last video here: http://sciinstitute.github.io/ShapeWorks/notebooks/getting-started-with-grooming-segmentations.html (this video matches what I just got running it on the release branch).

I guess we could go through the use case and notebook line by line and see where the grooming differs...

**Comment by jadie1** [comment]
Actually clicking and unclicking center doesn't cause a shift when you open the notebook distance transforms in studio, but it does for the use case distance transforms. So I believe the notebook is working as expected.
I'll start looking for where the use case differs.

**Comment by jadie1** [comment]
Scratch that the notebook and use case distance transforms have the same center/alignment issue. The distance transforms from the notebook are just smoother. 
@akenmorris do you think that the off centeredness may just be happening because there is a huge amount of variation in the dataset (size, orientation, radii, etc)?

**Comment by akenmorris** [comment]
They are being aligned with ICP?

**Comment by jadie1** [comment]
Yes, the reference is ellipsoid_14

**Comment by akenmorris** [comment]
So do you think it ends up aligning #9 along only one side?

I'm not sure how we should address this.  After ICP, should apply a center of mass operation?

@sheryjoe  any thoughts?  It seems wrong to capture this translation as part of the shape model.

**Comment by sheryjoe** [comment]
@akenmorris which translation? I can't see this in the video. Did you QC the center of mass alignmenet step? Also, number of ICP iterations could matter in some cases.

**Comment by akenmorris** [comment]
The ellipsoids are not centered, so necessarily the shape model captures the translation.

**Comment by sheryjoe** [comment]
> The ellipsoids are not centered, so necessarily the shape model captures the translation.

COM should center them.

**Comment by akenmorris** [comment]
> COM should center them.

I agree, but is ICP de-centering them to match one side?

**Comment by sheryjoe** [comment]
> > COM should center them.
> 
> I agree, but is ICP de-centering them to match one side?

The impact of ICP should be minimal wrt translation after COM, unless the reference is not really a median shape.

**Comment by akenmorris** [comment]
What's the status here?  Is anyone working on this?  I think there are two separate issues, grooming and then optimization parameters.

**Comment by sheryjoe** [comment]
@jadie1 @iyerkrithika21 any update?

**Comment by akenmorris** [comment]
Correct.  Ellipsoid 9 is centered before:

```
       """ Apply rigid alignment """
        ref = FindReferenceImage(comFiles)
        alignedFiles = applyRigidAlignment(groomDir + "aligned/segmentations", ref, comFiles)
```

And after, it is de-centered:

<img width="610" alt="image" src="https://user-images.githubusercontent.com/1693349/110503392-5ea35680-80b9-11eb-9690-158ada1b36af.png">

**Comment by iyerkrithika21** [comment]
@akenmorris when I try to optimize the ellipsoids in Studio without ICP, I get this message.
![image](https://user-images.githubusercontent.com/11266258/110507705-7e3c7e00-80bd-11eb-875a-5ad629fc5059.png)
I tried 3 different levels of padding - 10,30,50. But none of these helped.
And also in the python use case, I tried skipping the rigidAlignment step, and I get this bounding box error.
```
############## Cropping ##############
Traceback (most recent call last):
  File "RunUseCase.py", line 94, in <module>
    module.Run_Pipeline(args)
  File "/home/sci/iyerkrithika/ShapeWorks/Examples/Python/ellipsoid.py", line 96, in Run_Pipeline
    croppedFiles = applyCropping(groomDir + "cropped/segmentations", comFiles, comFiles)
  File "/home/sci/iyerkrithika/ShapeWorks/Examples/Python/GroomUtils.py", line 216, in applyCropping
    region = ImageUtils.boundingBox(bbDataList)
ValueError: Image sizes do not match (Output/ellipsoid/groomed/com_aligned/segmentations/ellipsoid_01.isores.center.com.nrrd)

```

This error doesn't make sense because the `ImageUtils.boundingBox()` should be able to give the largest bounding box.
@jadie1 do you think this could be related to the issue we saw in femur use case ?

**Comment by akenmorris** [comment]
@iyerkrithika21 , can you tell me more about how you're running these in studio?  If I had to guess based on the filenames in the corners, you perhaps imported the binary segmentations, skipped grooming and clicked run?  They need to be distance transforms.

**Comment by iyerkrithika21** [comment]
> 
> 
> @iyerkrithika21 , can you tell me more about how you're running these in studio? If I had to guess based on the filenames in the corners, you perhaps imported the binary segmentations, skipped grooming and clicked run? They need to be distance transforms.

I imported the segmentations, ran groom in studio. and then clicked optimize.

**Comment by jadie1** [comment]
Will be fixed for release via issue #1133

**Comment by akenmorris** [comment]
@iyerkrithika21 , I've added an issue for what you saw:

#1134
